### PR TITLE
Snapshot continued

### PR DESCRIPTION
**Pull request by alfrunes**
Follow-up PR for #480.

### CODE REVIEW

File: cli/snapshot.go [hunk 1]
```diff
@@ -85,32 +94,69 @@ func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
...
 		return nil, errors.Errorf(
-			"Data store (%s) is located on filesystem %s",
-			runOpts.dataStore, fsPath)
+			"data store (%s) is located on filesystem %s",
+			dataPath, fsPath)
+	}
+	rootDev, err := system.GetMountInfoFromDeviceID(rootDevID)
+	if err != nil {
+		return nil, err
+	} else if rootDev.FSType == "overlay" {
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply cli/snapshot.go]: It returns a device id with major number 0 which is, according to the documentation are non-device mounts. Perhaps it's better to detect a zero/non-block major device number?

File: cli/snapshot.go [hunk 2]
```diff
@@ -85,32 +94,69 @@ func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
...
-	return system.GetMountInfoFromDeviceID(rootDevID)
+	return rootDev, nil
 }
 
 func prepareOutStream(out io.WriteCloser, compression string) (io.WriteCloser, error) {
+
+	var pSize uint64
+
+	if f, ok := out.(*os.File); ok {
+		pSize = uint64(system.GetPipeSize(int(f.Fd())))
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply cli/snapshot.go]: It's just a way of detecting if the pipe is read on the other side - It's needed because this will prohibit the progress bar from cluttering possible prompts from other process(es).

[the remark this answers is not in the document]
Comment (alfrunes) [reply cli/snapshot.go]: I create a start condition for the progress bar print function based on the pipe (or actually input) buffer size.

File: cli/snapshot.go [hunk 3]
```diff
@@ -119,98 +165,102 @@ func prepareOutStream(out io.WriteCloser, compression string) (io.WriteCloser, e
...
+		}
+		if err != nil {
 			log.Warnf("Failed to freeze filesystem on %s: %s",
-				rootDev.MountPoint, err.Error())
+				rootDev.MountSource, err.Error())
 			log.Warn("The snapshot might become invalid.")
-			abortChan <- struct{}{} // abort handler
-			signal.Stop(sigChan)
+			close(abortChan)
+			stopFreezeHandler(sigChan, abortChan)
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply cli/snapshot.go]: If it's `system.FreezeFS` that caused the error the `freezeHandler` will be hanging doing nothing.

File: cli/snapshot.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kacf) [reply cli/snapshot.go]: Yes, that sounds exactly like what we want.

File: cli/snapshot.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kacf) [reply cli/snapshot.go]: Oh, I missed the extra `err` assignment there. But in that case you have already called `defer stopFreezeHandler(sigChan, abortChan)`, so I still think this is an unnecessary call.

File: cli/snapshot.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kacf) [reply cli/snapshot.go]: Alright, that makes sense. Please add some comments about that though. The code doesn't make it clear why this is needed.

File: cli/snapshot.go [hunk 4]
```diff
@@ -32,28 +32,37 @@ import (
...
+	// bufferSize for the Copy function
+	bufferSize = 32 * 1024
 )
 
 var (
 	errWatchDogExpired = fmt.Errorf("watchdog timer expired")
+
+	// pbCondition is used to determine when the progress bar can start
+	// printing. This value is nil if the snapshot is not dumped to a pipe.
+	pbCondition utils.StartCondition
```
Comment (kacf) [inline cli/snapshot.go]: Could we find a way to pass this in the arguments instead? Or in a struct? I don't think this should be global.

**Comment by kacf** [comment]
At least all the mender-qa tests [passed now](https://gitlab.com/Northern.tech/Mender/mender-qa/pipelines/113013402)! Only unit tests to fix.

**Comment by kacf** [comment]
FYI, I added [a test case](https://github.com/kacf/meta-mender/commits/snapshot) for the block device argument to `-fs-path`, which currently fails locally. However, I prefer to get the existing tests passing for now, so I have not added it to https://github.com/mendersoftware/meta-mender/pull/897. We can look at it after this PR has passed and been merged into [snapshot-no-ssh](https://github.com/mendersoftware/mender/commits/snapshot-no-ssh).

File: cli/snapshot.go [hunk 5]
```diff
@@ -76,199 +86,231 @@ func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
...
-	var rootStat unix.Stat_t
-	if err := unix.Fstat(fd, &stdoutStat); err != nil {
-		return errors.Wrap(err, "Unable to stat output file")
-	}
-	if err := unix.Stat(ctx.String("fs-path"), &rootStat); err != nil {
-		return errors.Wrap(err, "Unable to stat root filesystem")
-	}
-	if stdoutStat.Dev == rootStat.Dev {
-		return errors.New(
-			"Dumping the filesystem to itself is not permitted")
```
Comment (kacf) [inline cli/snapshot.go]: Are you relying completely on the freeze timer now? This message is being tested for, so we need to change the test, in that case.

File: cli/snapshot.go [hunk 6]
```diff
@@ -32,28 +32,37 @@ import (
 )
 
 const (
-	errMsgDataPartF = "Device-local data is stored on the partition " +
-		"being snapshotted: %s. The recommended approach is to have " +
-		"a separate data-partition mounted on \"/data\" and add a " +
+	// Error messages //
+	errMsgDataPartFmt = "Device-local data is stored on the rootfs " +
+		"partition: %s. The recommended approach is to have  a " +
+		"separate data-partition mounted on \"/data\" and add a " +
```
[the remark this answers is not in the document]
Comment (kacf) [reply cli/snapshot.go]: Still no change here.

File: system/ioctl.go [hunk 7]
```diff
@@ -136,25 +135,20 @@ func GetMountInfoFromDeviceID(devID [2]uint32) (*MountInfo, error) {
...
 // the inode at path.
 func GetDeviceIDFromPath(path string) ([2]uint32, error) {
-	var stat unix.Stat_t
+	var stat stat
 
-	if err := unix.Stat(path, &stat); err != nil {
-		return [2]uint32{^uint32(0), ^uint32(0)}, err
+	if err := sys.Stat(path, &stat); err != nil {
+		return [2]uint32{^uint32(0), ^uint32(0)}, errors.Wrapf(err,
+			"stat %s", path)
```
Comment (kacf) [inline system/ioctl.go]: The code I want to comment on is a little bit further down, but it is from an earlier commit, so I can't put the comment there. However, following our discussion from yesterday, I think we may want to change this code a bit:

```go
	switch devType {
	// If path refers to a special file (e.g. device file under /dev), then
	// st_dev refers to the device number of the mounted devfs. The device
	// number for a special file is under the st_rdev property, ref stat(2).
	case unix.S_IFBLK,
		unix.S_IFCHR,
		unix.S_IFIFO,
		unix.S_IFSOCK:
		return [2]uint32{
			unix.Major(stat.Rdev), unix.Minor(stat.Rdev)}, nil
	// If path refers to a regular file, then st_dev gives the device number
	// of the underlying block device.
	case unix.S_IFDIR,
		unix.S_IFREG,
		unix.S_IFLNK:
		return [2]uint32{
			unix.Major(stat.Dev), unix.Minor(stat.Dev)}, nil
	}
```

1. I think the first case should only cover `S_IFBLK`. No other device is appropriate for taking the `Rdev`. I mean, you *can* do it for a `S_IFCHR` device as well, but it's not valid for snapshots.
2. I would opt for being somewhat restrictive and make the second case only cover `S_IFDIR`, with a `default` that covers the rest with an error message (`S_IFLNK` is implicitly covered by `S_IFDIR`, because you are using `Stat`, not `Lstat`). I think this is less surprising. For example, consider the case where a user tries to do this:
   ```bash
   mender snapshot dump /home/user/my-test-file.fs
   ```
   This will dump the filesystem which the file is on, but it looks more like it should take a snapshot of that file. Instead I think an error message here is fine, since dumping a file is kind of useless (you can just copy it).

So in short, only block devices and directories would be valid arguments to `--source`. What do you think?

Comment (alfrunes) [reply system/ioctl.go]: Yes and no. I agree that the source-flag can be confusing in the file-case. However, this function is in the system package, and therefore should not be tied to requirements in the cli package. If there's a strong sentiment for discarding file and non-device special files from the source flag, then I think the cli package should do the stat and mode-check.
On a different note, for systems running overlay passing a directory to `--source` will wail, but a file will result in the device storing the file being resolved.

File: cli/snapshot.go (hunk 5 above)
Comment (alfrunes) [reply cli/snapshot.go]: Yes, I don't think we should treat this redirection case any different from pipes. So just let the device freeze until the timer expires. That's the punishment of incorrect usage :smiling_imp:

Comment (kacf) [reply cli/snapshot.go]: Hehe ok, I will adjust the test. :-)

File: system/ioctl.go (hunk 7 above)
Comment (kacf) [reply system/ioctl.go]: > However, this function is in the system package, and therefore should not be tied to requirements in the cli package.

Yes, but we put it there, we should not let that rule functionality. After all it's the user experience that counts. I agree that it might fit better in the cli package for this use case, so we can just move it.

> On a different note, for systems running overlay passing a directory to --source will wail, but a file will result in the device storing the file being resolved.

Yes, but this is very unintuitive, and I think I would prefer an error message in this case. If you try to reference a file like this, you don't even know which device you will snapshot. Will you get the the base branch? Or maybe the COW branch, which is nearly empty? It can even change from one invocation to the next if you write to the file in between. I think requiring directories or block devices is cleaner and less confusing.

File: cli/snapshot.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alfrunes) [reply cli/snapshot.go]: Sorry, just forgot it the first time...

**Comment by kacf** [comment]
@alfrunes: You probably won't like it, but my proposed fix for the hanging is [here](https://github.com/mendersoftware/mender/pull/493). Yes, it's much uglier, but we need something that works now. And it's not worse than it already was.

This passed everything locally, and hopefully will pass on Gitlab as well. If it does, and you want to keep working on the progress bar, we can reconsider it in a different PR.

**Comment by alfrunes** [comment]
> @alfrunes: You probably won't like it, but my proposed fix for the hanging is [here](https://github.com/mendersoftware/mender/pull/493). Yes, it's much uglier, but we need something that works now. And it's not worse than it already was.

It's alright, the progress-bar has caused way too much hassle at this point. It's probably about time we get rid of it for now.

**Comment by alfrunes** [comment]
> This passed everything locally, and hopefully will pass on Gitlab as well. If it does, and you want to keep working on the progress bar, we can reconsider it in a different PR.

There's still a possible deadlock that can occur when the timer expires in `cli/snapshot.go:freezeHandler`, it's quite rare as it only happen upon a well timed context switch. I provided a fix for this case in my latest commit.

**Comment by kacf** [comment]
> It's alright, the progress-bar has caused way too much hassle at this point. It's probably about time we get rid of it for now.

If nothing else, it did put up a good fight! :-)

> There's still a possible deadlock that can occur when the timer expires in `cli/snapshot.go:freezeHandler`, it's quite rare as it only happen upon a well timed context switch. I provided a fix for this case in my latest commit.

Unfortunately I could not get that to work either. I decided to try to rework the freeze handler into [a more common request/response pattern](https://github.com/mendersoftware/mender/pull/493/commits/cc399319b6fb899a72b413e8d2be5263473fdb79). This passed both locally and on Gitlab, but I was running it manually, so let's see how the the final run does.

**Comment by kacf** [comment]
Superseded by #493.
